PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc

```diff
@@ -1,92 +1,400 @@
+        reco::TrackRef tkref = reco::TrackRef(gt_h, igt);
+        pat::PackedCandidateRef pcref = (*gt2pc)[tkref];
+        pat::PackedCandidateRef ltref = (*gt2lt)[tkref];
+        const pat::PackedCandidate & pfCand = *(pcref.get());
+        const pat::PackedCandidate & lostTrack = *(ltref.get());
+
+        // Determine if this general track is associated with anything in packedPFCandidates or lostTracks
+        // Sometimes, a track gets associated w/ a neutral pfCand.
+        // In this case, ignore the pfCand and take from lostTracks
+        bool isInPackedCands = (pcref.isNonnull() && pcref.id()==pc_h.id() && pfCand.charge()!=0);
+        bool isInLostTracks  = (ltref.isNonnull() && ltref.id()==lt_h.id());
+
+        LorentzVector p4;
+        pat::PackedCandidateRef refToCand;
+        int pdgId, charge, fromPV;
+        float dz, dxy, dzError, dxyError;
+        int pfCandInd; //to avoid counting packedPFCands in their own isolation
 
-        if(pf_it->p4().pt() < pT_cut)
+        // get the four-momentum and charge
+        if(isInPackedCands){
+            p4        = pfCand.p4();
+            charge    = pfCand.charge();
+            pfCandInd = pcref.key();
+        }else if(isInLostTracks){
+            p4        = lostTrack.p4();
+            charge    = lostTrack.charge();
+            pfCandInd = -1;
+        }else{
+            double m = 0.13957018; //assume pion mass
+            double E = sqrt(m*m + gentk.p()*gentk.p());
+            p4.SetPxPyPzE(gentk.px(), gentk.py(), gentk.pz(), E);
+            charge = gentk.charge();
+            pfCandInd = -1;
+        }
+
+        if(p4.pt() < pT_cut_)
+            continue;
+        if(charge == 0)
             continue;
 
-        if(pf_it->fromPV()<=1 && fabs(pf_it->dz()) > dZ_cut)
+        // get the isolation of the track
+        pat::PFIsolation isolationDR03;
+        pat::PFIsolation miniIso;
+        getIsolation(p4, pc, pfCandInd, isolationDR03, miniIso);
+        
+        // isolation cut
+        if( p4.pt() < pT_cut_noIso_ && 
+            !(isolationDR03.chargedHadronIso() < absIso_cut_ ||
+              isolationDR03.chargedHadronIso()/p4.pt() < relIso_cut_ ||
+              miniIso.chargedHadronIso()/p4.pt() < miniRelIso_cut_))
             continue;
 
-        if(pf_it->charge() == 0)
+        // get the rest after the pt/iso cuts. Saves some runtime
+        if(isInPackedCands){
+            pdgId     = pfCand.pdgId();
+            dz        = pfCand.dz();
+            dxy       = pfCand.dxy();
+            dzError   = pfCand.dzError();
+            dxyError  = pfCand.dxyError();
+            fromPV    = pfCand.fromPV();
+            refToCand = pcref;
+        }else if(isInLostTracks){
+            pdgId     = 0;
+            dz        = lostTrack.dz();
+            dxy       = lostTrack.dxy();
+            dzError   = lostTrack.dzError();
+            dxyError  = lostTrack.dxyError();
+            fromPV    = -1;
```

Comment:
... actually, the origin is somewhat clearly defined by refToCand. So, no extra enums are needed for that.